### ISSUE TITLE
[Flutter Web] authCallback not call

### ISSUE THREAD

**Issue by buimanhhungit** [milestonetg/msal-js-dart#8]
Pls check for me : handleRedirectCallback not callback and loginpopup not working when publish  to websv but working when debug.
Im using flutter.

![Screen Shot 2020-07-02 at 10 00 14](https://user-images.githubusercontent.com/51559988/86311558-db3c8600-bc4a-11ea-8c3c-7acd541ae9cb.png)

![Screen Shot 2020-07-02 at 10 01 30](https://user-images.githubusercontent.com/51559988/86311648-0cb55180-bc4b-11ea-9012-559644e8f33f.png)

![Screen Shot 2020-07-02 at 10 01 41](https://user-images.githubusercontent.com/51559988/86311651-0e7f1500-bc4b-11ea-883a-24d1f197929b.png)

![Screen Shot 2020-07-02 at 10 07 44](https://user-images.githubusercontent.com/51559988/86312027-e5ab4f80-bc4b-11ea-8437-ad9de636ca8f.png)

**Comment by Francessco121** [comment]
Not immediately sure why that would be breaking in release mode but here's some things to try:
- Try using msal.js 1.2.1, I haven't updated the library for 1.3.x yet (shouldn't be an issue but might help).
- Setup an MSAL logger if you haven't already and set its level to verbose. What gets logged?
- Make sure `handleRedirectCallback` is getting called, that function is what completes a redirect login.

Here's an example of setting up the logger:
```dart
final logger = Logger(loggerCallback,
  LoggerOptions()
    ..level = LogLevel.verbose
);

final userAgentApplication = UserAgentApplication(
  Configuration()
    ..system = (SystemOptions()
      ..logger = logger
    )
);

// ...

void loggerCallback(LogLevel level, String message, bool containsPii) {
  print('[$level] $message');
}
```

It sounds like MSAL isn't loading or being setup in release mode. Try the logger and see if MSAL logs any errors or warnings.

**Comment by buimanhhungit** [comment]
I tried it with msal.js 1.2.1 but:

1. loginRedirect
   I have successfully logged in Microsoft and your library not have log, not callback.

2. loginPopup
  With debug: it's ok
![Screen Shot 2020-07-02 at 13 50 21](https://user-images.githubusercontent.com/51559988/86325904-18b10b80-bc6b-11ea-9bd0-1e4f1b266079.png)


when publish to web server: it's not ok and your library not have log.

![Screen Shot 2020-07-02 at 13 37 42](https://user-images.githubusercontent.com/51559988/86326763-5febcc00-bc6c-11ea-8ef5-883c1d7b8df6.png)

![Screen Shot 2020-07-02 at 13 38 39](https://user-images.githubusercontent.com/51559988/86326732-57939100-bc6c-11ea-9e0c-e888d8a602bd.png)

![Screen Shot 2020-07-02 at 13 39 44](https://user-images.githubusercontent.com/51559988/86326705-4cd8fc00-bc6c-11ea-9f37-c489b0839740.png)

**Comment by Francessco121** [comment]
Ok, based on what you're giving me here it seems like there's two different issues going on. 

First, it looks like you're only calling `handleRedirectCallback` when you want to initiate a `loginRedirect`, which isn't entirely correct. `loginRedirect` doesn't use the callback you give to `handleRedirectCallback`. Instead, `handleRedirectCallback` should be called when your app first loads usually right after you instantiate your `UserAgentApplication` as it is what gets the token off of the url after a redirect.

Second, it really seems like msal.js isn't even loading at all in your deployed environment. Can you confirm in your browsers network tab on your deployed app that the msal.js script is loading correctly?

If there's any way you could provide a more complete sample of your code, that would help a lot. I don't really have enough information here to try to replicate the issue you're having.

**Comment by buimanhhungit** [comment]
Does msal.js need to be installed on web server?
or just put in index.html?

**Comment by Francessco121** [comment]
It needs to be referenced in `index.html`. That can be a reference to the CDN, or to a local file in your `web` folder. If it's not using a CDN, make sure the file exists in the deployed build (there was a Flutter bug a while back that caused files other than `index.html` under `web` to not get copied to the build output, but supposedly that was fixed).

**Comment by buimanhhungit** [comment]
it was working at debug in flutter. But i deploy to webser then I have successfully logged in Microsoft and your library not have log, not runing.

**Comment by buimanhhungit** [comment]
Can i video call for you?

Vào 08:31, T.6, 3 Th7, 2020 Ethan Lafrenais <notifications@github.com> đã
viết:

**Comment by Francessco121** [comment]
Sorry, a video call isn't gonna work for me.

This really isn't enough information for me to go off of. It seems like this is an environment or MSAL issue.
Just so I'm clear, in Flutter Web debug mode the whole login flow works, and when you deploy it you're able to login but after the redirect things stop working?

Please check your browser console for errors in your _deployed_ app, if you could post most of the console logs here that'd be really helpful to narrow down the actual error that's happening. Especially what gets logged after the redirect.

If you could also post the complete code you have interacting with MSAL, that would also be super helpful as I might be able to recreate the issue you're having.

**Comment by buimanhhungit** [comment]
yes, redirect things stop working with loginpopup. 

this is console logs:
![Screen Shot 2020-07-03 at 09 04 04](https://user-images.githubusercontent.com/51559988/86424273-4c903d80-bd0c-11ea-8df2-07257545825e.png)

**Comment by Francessco121** [comment]
Ok, so that's a step in the right direction. Unfortunately, this is all minified code.

Can you deploy your app in `profile` mode, so Flutter doesn't minify the code, and post the error again?
The command should be `flutter build web --profile`.

The unminified stack trace should narrow down this error.

**Comment by buimanhhungit** [comment]
i built with command flutter build web --profile but nothing change. ' method not found: 'call'...

![Screen Shot 2020-07-03 at 09 38 29](https://user-images.githubusercontent.com/51559988/86426133-1bfed280-bd11-11ea-9880-15979e8022d8.png)

**Comment by Francessco121** [comment]
That helps a lot actually, it shows the function names now in the stack trace.

I don't think that's the full stack trace though, it doesn't show your code or code from this package. Could you paste the full thing here? Usually Dart web errors will print out two fairly large stack traces in the console, the Dart stack trace and the JavaScript stack trace.

If you could also narrow down which call to this package throws this error, that would be super helpful.

**Comment by buimanhhungit** [comment]
![Screen Shot 2020-07-03 at 09 59 50](https://user-images.githubusercontent.com/51559988/86427270-fa531a80-bd13-11ea-8b1e-2be2c6903ded.png)

in dart code, i just call to loginPopup like to your example and test...

**Comment by Francessco121** [comment]
Your index.html file looks fine.

That error that you posted earlier however isn't the full error, and that full error is most likely the cause of the redirect callback not being invoked.

Could you please copy/paste the full error here?

I can't even begin to determine whether this is a bug in this package without the actual error occurring.

**Comment by buimanhhungit** [comment]
how to view full error ?

**Comment by Francessco121** [comment]
I guess that depends on what browser you're using, but you should just be able to scroll down further. Just select all of the text in the console and paste it here.

If that is actually the full and only error you're getting, that unfortunately doesn't narrow anything down... You'll need to insert some print statements to figure out how far your code gets before it throws that error in that case.

**Comment by Francessco121** [comment]
I've managed to reproduce the error you're having and are trying to figure out where/why this is happening. It seems that Flutter changed some things recently and broke this package's compatibility with it.

**Comment by buimanhhungit** [comment]
yes. tks

**Comment by buimanhhungit** [comment]
My guess is it's related to this allowInterop call?

**Comment by Francessco121** [comment]
Yeah, I narrowed the issue down to a Dart SDK bug with the `dart2js` compiler, sort of. I can confirm that this issue does not happen when using `dart2js` normally, but the way or version Flutter uses has a bug that breaks JavaScript callbacks, specifically `allowInterop`.

It's really weird, you're supposed to wrap functions sent to JavaScript in `allowInterop`, otherwise it won't work when compiled to `dart2js`. And it's true if you're not using Flutter, however the opposite seems to be true in Flutter where using `allowInterop` breaks it completely.

I came across this SDK issue that was supposedly resolved a while back https://github.com/dart-lang/sdk/issues/37576. However, trying the suggested fix here actually works for me. I'd really rather not add a dependency on `package:js` given that this package isn't using it and that this is definitely a Flutter bug, but we'll see.

If you want a work-around for the time being:
- Add a dependency on `package:js`
- In your `main.dart` file (or really any file), add a library directive like `library main;` and annotate it with `@JS()`

Ex:
```dart
@JS()
library main;

import 'package:js/js.dart';
```
